src/dimensions.jl

```diff
@@ -0,0 +1,157 @@
+
+"""
+    has_dimnames(x) -> Bool
+
+Returns `true` if `x` has names for each dimension.
+"""
+@inline has_dimnames(x) = has_dimnames(typeof(x))
+function has_dimnames(::Type{T}) where {T}
+    if parent_type(T) <: T
+        return false
+    else
+        return has_dimnames(parent_type(T))
+    end
+end
+
+"""
+    dimnames(x) -> Tuple{Vararg{Symbol}}
+    dimnames(x, d) -> Symbol
+
+Return the names of the dimensions for `x`.
+"""
+@inline dimnames(x) = dimnames(typeof(x))
+@inline dimnames(x, i::Integer) = dimnames(typeof(x), i)
+@inline dimnames(::Type{T}, d::Integer) where {T} = getfield(dimnames(T), to_dims(T, d))
+@generated function dimnames(::Type{T}) where {T}
+    if parent_type(T) <: T
+        return ntuple(i -> Symbol(:dim_, i), Val(ndims(T)))
```

Comment:
As you know, I've had mixed experience with having default names (e.g., https://github.com/JuliaImages/ImageCore.jl/issues/112). Recently, though, you've played around with this a lot more than I have. Is your sense that this really does make life better?

Reply:
Would it make sense to do what NamedDims.jl does and return `:_` for unnamed dimensions? Even if we don't have default names for arrays that return false on `has_dimnames` then we have to come up with something for the transpose/adjoint of a vector or when creating a view of an array with extra slices (e.g., `ones(2,2)[:,:,:,:]`).

Another slightly related problem is this could lead to multiple dimensions named `:_`. It makes more sense to me to have dimensions with strictly unique names like a `NamedTuple`, but perhaps that's not the right way of thinking about it.

Reply:
I confess I haven't had time to play with NamedDims, although in the (hopefully not too) long run I definitely want to transition JuliaImages to something that is actively maintained. Using a default, non-unique name seems like a pretty good solution to me. My own experience suggests that giving things unique but meaningless names doesn't seem likely to be very helpful, and might be hard to distinguish unambiguously from intentional names. (What if users deliberately choose `dim_1`, `dim_2`, etc? That would spoil any attempt to detect whether the names are meaningful or just default.)

Reply:
Perhaps it's worth saying that the default name `:_` is a wild-card, which matches anything. As you say, generated names `:dim_1 ` etc. would probably need to be tagged as being 2nd-class somehow, so that clashes aren't taken as seriously as clashes of meaningful names.

I agree that the NamedTuple analogy is appealing. But I was persuaded that allowing even meaningful names to repeat is natural, as they arise from things like `m = v .+ v'`, or correlation matrices. In such cases `sum(m, dims=:i)` currently picks the first matching dimension, and picking one certainly seems like the right behaviour. (I'm not sure this is tested & guaranteed.) `m[i=1]` fixes both, but people like to say it's undefined, and should perhaps be made an error. The alternative would be some mechanism for always adding primes in such cases, I'm not sure this has been much explored. 

(Have not thought much about the broader proposal here, and whether this should live here not in NameDims.)

Reply:
Okay. I'm convinced that that the old version with `:dim_i` naming only complicates the problem I was trying to solve. I think the wild card (`:_`) also helps avoid defining any formal rules about repeated dimension names here, and pushes that off as an implementation detail for packages like `NamedDims.jl`.